### PR TITLE
Added information about security aspects in Ephemeral Containers

### PR DESCRIPTION
**Pull request by deepak4566**
Solves #42319 
Added information about security aspects of ephemeral containers

### CODE REVIEW

**Review by sftim — COMMENTED** [review]
Thanks for the PR. Please update it to be in line with our [style guide](https://kubernetes.io/docs/contribute/style/style-guide/).

You can [preview the website locally](https://kubernetes.io/docs/contribute/new-content/open-a-pr/#preview-locally) to see how you're doing.

File: content/en/docs/concepts/workloads/pods/ephemeral-containers.md [hunk 1]
```diff
@@ -76,3 +76,195 @@ you can view processes in other containers.
...
+      seLinux:
+          rule: RunAsAny
+            supplementalGroups:
+                rule: RunAsAny
+                  runAsUser:
+                      rule: RunAsAny
+                        fsGroup:
+                            rule: RunAsAny
+                              volumes:
+                                - '*'
```
Comment (windsonsea) [inline content/en/docs/concepts/workloads/pods/ephemeral-containers.md:143]:
```suggestion
```yaml
apiVersion: policy/v1beta1
kind: PodSecurityPolicy
metadata:
  name: default
  spec:
    privileged: false
      seLinux:
          rule: RunAsAny
            supplementalGroups:
                rule: RunAsAny
                  runAsUser:
                      rule: RunAsAny
                        fsGroup:
                            rule: RunAsAny
                              volumes:
                                - '*'
```
```

File: content/en/docs/concepts/workloads/pods/ephemeral-containers.md [hunk 2]
```diff
@@ -76,3 +76,195 @@ you can view processes in other containers.
...
+          - pods/attach
+            - pods/exec
+              - pods/portforward
+                - pods/proxy
+                  verbs:
+                    - create
+                      - delete
+                        - deletecollection
+                          - patch
+                            - update
```
Comment (windsonsea) [inline content/en/docs/concepts/workloads/pods/ephemeral-containers.md:176]:
```suggestion
```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: edit
  rules:
  - apiGroups:
    - ""
      resources:
        - pods
          - pods/attach
            - pods/exec
              - pods/portforward
                - pods/proxy
                  verbs:
                    - create
                    - delete
                    - deletecollection
                    - patch
                    - update
```
```
Not sure the indentations correct or not. Seems weird.

File: content/en/docs/concepts/workloads/pods/ephemeral-containers.md [hunk 3]
```diff
@@ -76,3 +76,195 @@ you can view processes in other containers.
...
+
+Ephemeral containers are created using a special ephemeralcontainers handler in the API rather than by adding them directly to pod.spec, so it's not possible to add an ephemeral container using kubectl edit.
+
+The simplest method to add an ephemeral container with a security context to a Pod is to use the Go client. A couple of lines of code can add a new ephemeral container running as privileged or use any other security context setting which is to your liking.
+
+package main
```
Comment (windsonsea) [inline content/en/docs/concepts/workloads/pods/ephemeral-containers.md:186]: This kind of code block is hard to read.

File: content/en/docs/concepts/workloads/pods/ephemeral-containers.md [hunk 4]
```diff
@@ -76,3 +76,195 @@ you can view processes in other containers.
 
 * Learn how to [debug pods using ephemeral containers](/docs/tasks/debug/debug-application/debug-running-pod/#ephemeral-container).
 
+### Kubernetes Ephemeral Container Security
+Attempting to debug a Pod and realizing that you can't install curl due to security settings has to be a meme at this point. Good security practices are always nice but it often comes at the cost of usability. To the point where some may even solve this problem by installing debug tools into their production images. Shudders.
```
Comment (tengqm) [inline content/en/docs/concepts/workloads/pods/ephemeral-containers.md:80]: Please manually wrap long lines like this.

Long lines are difficult to comment, to translate, to track changes.

File: content/en/docs/concepts/workloads/pods/ephemeral-containers.md [hunk 5]
```diff
@@ -76,3 +76,195 @@ you can view processes in other containers.
...
+
+Kubernetes has introduced a new concept called ephemeral containers to deal with this problem. Ephemeral containers are temporary containers that can be attached after a Pod has been created. Rejoice! We can now attach a temporary container with all the tools which we desire. While the applications container may have "annoying security features" like a read only file system the ephemeral container can enjoy all the freedom which writing files entails. I love this feature so I need to upgrade my cluster immediately
+
+
+## Digging Deeper
+
+Now that we have the new feature we can start a ephemeral container in any Pod we like  
+
+kubectl run ephemeral-demo --image=k8s.gcr.io/pause:3.1 --restart=Never
+kubectl debug -it ephemeral-demo --image=busybox:1.28
```
Comment (tengqm) [inline content/en/docs/concepts/workloads/pods/ephemeral-containers.md:90]:
````suggestion
```shell
kubectl run ephemeral-demo --image=k8s.gcr.io/pause:3.1 --restart=Never
kubectl debug -it ephemeral-demo --image=busybox:1.28
```
````

File: content/en/docs/concepts/workloads/pods/ephemeral-containers.md [hunk 6]
```diff
@@ -76,3 +76,195 @@ you can view processes in other containers.
...
+
+We get a shell and life is now much simpler, but wait a minute. This post is not about how to use ephemeral containers, there are enough of those already, but rather the security implications of enabling ephemeral containers. Let's have a look at the YAML for the Pod that we created the ephemeral container in.
```
Comment (tengqm) [inline content/en/docs/concepts/workloads/pods/ephemeral-containers.md:92]: Please carefully read the documentation style in the contribution guide.
Please avoid using 'we', avoid using word like 'simpler', 'simple', 'easy', and so on.
Please use simple precise English. For example, ", there are enough of those already, " isn't something you want to talk about. It is irrelevant to the topic of this page.
"We get a shell and life is now much simpler ... ", no, we don't appreciate this wordy statements.

File: content/en/docs/concepts/workloads/pods/ephemeral-containers.md [hunk 7]
```diff
@@ -76,3 +76,195 @@ you can view processes in other containers.
...
+## OPA Gatekeeper
+
+OPA Gatekeeper does not require any code changes as all of its policies are written in rego. It's sub project Gateekper Library does however have to be updated. The library contains an implementation of the common Pod Security Policies. This includes policies like not allowing containers in privileged mode. The issue with the all of the policies is that they currently only check containers specified in initContainers and containers, analyze the following rego as an example.
+
+The good news is that this is a pretty easy fix, the bad news is that it requires end users to update the policies pulled from the library.
+
+## Kyverno
+
+Kyverno seems to have resolved the issues faster. Compared to OPA Gatekeeper however it did require a small code change which means that version 1.5.3 or later is needed to write policies for ephemeral containers. They have also updated their policy library to include checking ephemeral containers. Kyverno has done a great job solving these issues quickly. It does still require end users to update however.
+
```
Comment (tengqm) [inline content/en/docs/concepts/workloads/pods/ephemeral-containers.md:121]: No. We don't talk about community solutions that are out of the kubernetes project.
There are simply too many of them, we cannot afford to cover or to track.

File: content/en/docs/concepts/workloads/pods/ephemeral-containers.md [hunk 8]
```diff
@@ -76,3 +76,195 @@ you can view processes in other containers.
...
```
Comment (tengqm) [inline content/en/docs/concepts/workloads/pods/ephemeral-containers.md:269]: This is a good blog article if you want to publish it somewhere else.
However, the taste of this page is very different from what we host on the k8s.io/docs website.

Comment (deepak4566) [reply content/en/docs/concepts/workloads/pods/ephemeral-containers.md:269]: @tengqm @windsonsea  thanks for ur doc review it's my first documentation for an open source project so I didn't know the guidelines so I may made some mistakes and I  will follow ur suggestions for the  changes proposed

File: content/en/docs/concepts/workloads/pods/ephemeral-containers.md [hunk 9]
```diff
@@ -76,3 +76,195 @@ you can view processes in other containers.
...
+  spec:
+    ...
+      ephemeralContainers:
+        - name: debugger-r59b7
+            image: busybox:1.28
+                imagePullPolicy: IfNotPresent
+                    stdin: true
+                        terminationMessagePath: /dev/termination-log
+                            terminationMessagePolicy: File
+                                tty: true
```
[the remark this answers is not in the document]
Comment (tengqm) [reply content/en/docs/concepts/workloads/pods/ephemeral-containers.md:110]: The indentation is totally a mess.
Please check.

File: content/en/docs/concepts/workloads/pods/ephemeral-containers.md [hunk 10]
```diff
@@ -76,3 +76,208 @@ you can view processes in other containers.
...
+  name: ephemeral-demo
+spec:
+  ...
+  ephemeralContainers:
+  - name: debugger-r59b7
+    image: busybox:1.28
+    imagePullPolicy: IfNotPresent
+    stdin: true
+    terminationMessagePath: /dev/termination-log
+    terminationMessagePolicy: File
```
Comment (tengqm) [inline content/en/docs/concepts/workloads/pods/ephemeral-containers.md:109]: Why do we need to explicitly specify these attributes?

File: content/en/docs/concepts/workloads/pods/ephemeral-containers.md [hunk 11]
```diff
@@ -76,3 +76,208 @@ you can view processes in other containers.
...
+    tty: true
+```
+                            
+Interesting, there is a new field called ephemeralContainers in the Pod definition. This new field contains a list of containers similar to initContainers and containers. It is not identical as there are certain options which are not available, refer to the API documentation for more information. It does however allow configuration of the container security context, which could in theory allow a bad actor to escalate the container's privileges. This should not affect those of us who use a policy enforcement tool right? The answer is yes and no depending on the tool and version that is being used. It also depends on if you are using policies from the project's library or policies developed in house.
+
```
Comment (tengqm) [inline content/en/docs/concepts/workloads/pods/ephemeral-containers.md:114]: Please save the irrelevant statements such as "interesting".
The statement about "new" is warranted to decay pretty soon.
Please try use backticks for field names for readability.

```
The `ephemeralContainers` field in the spec of a Pod ...
```

Comment (tengqm) [reply content/en/docs/concepts/workloads/pods/ephemeral-containers.md:114]: "This should not affect those of us who use a policy enforcement tool right? The answer is yes and no depending on the tool and version that is being used."

Please just state the facts in plain English.

File: content/en/docs/concepts/workloads/pods/ephemeral-containers.md [hunk 12]
```diff
@@ -76,3 +76,208 @@ you can view processes in other containers.
...
+          - pods/attach
+            - pods/exec
+              - pods/portforward
+                - pods/proxy
+                  verbs:
+                    - create
+                    - delete
+                    - deletecollection
+                    - patch
+                    - update
```
Comment (tengqm) [inline content/en/docs/concepts/workloads/pods/ephemeral-containers.md:182]: The indentation of the snippet is a complete mess.

File: content/en/docs/concepts/workloads/pods/ephemeral-containers.md [hunk 13]
```diff
@@ -76,3 +76,208 @@ you can view processes in other containers.
...
+
+Ephemeral containers are created using a special ephemeralcontainers handler in the API rather than by adding them directly to pod.spec, so it's not possible to add an ephemeral container using kubectl edit.
+
+The simplest method to add an ephemeral container with a security context to a Pod is to use the Go client. A couple of lines of code can add a new ephemeral container running as privileged or use any other security context setting which is to your liking
+```
+package main
+
+import (
```
Comment (tengqm) [inline content/en/docs/concepts/workloads/pods/ephemeral-containers.md:195]: Is this Go program helpful to understand the concept of ephemeral containers?
I don't get the point.

**Comment by deepak4566** [comment]
Hey sir @tengqm can I expect any update this pr? Any other changes need to be done??or need to approve this?

**Comment by deepak4566** [comment]
Hey sir @tengqm any update regarding this pr? Or any active changes need to be done?

File: content/en/docs/concepts/workloads/pods/ephemeral-containers.md [hunk 14]
```diff
@@ -76,3 +76,267 @@ you can view processes in other containers.
...
+are enforced in accordance to your security practices. You may have 
+been aware of the API changes and taken the correct precautionary 
+steps. Or you just updated Kyverno and it's policies out of pure
+happenstance. Either way it is good to trust but verify that
+it is not for example possible to create a privileged ephemeral 
+container. Annoyingly the debug command does not expose any options
+to set any security context configuration, so we need another option.
+Ephemeral containers cannot be defined in a Pod when it is created 
+and it can neither be added with an update. We need some other
+method to create these specific ephemeral containers.
```
Comment (tengqm) [inline content/en/docs/concepts/workloads/pods/ephemeral-containers.md:236]: The above paragraph still reads like a blog post.

Comment (sftim) [reply content/en/docs/concepts/workloads/pods/ephemeral-containers.md:236]: (if you want to write a blog article instead @deepak4566, that's an option you have)

File: content/en/docs/concepts/workloads/pods/ephemeral-containers.md [hunk 15]
```diff
@@ -76,3 +76,261 @@ you can view processes in other containers.
...
+  ...
+  ephemeralContainers:
+  - name: debugger-r59b7
+    image: busybox:1.28
+    imagePullPolicy: IfNotPresent
+    stdin: true
+    tty: true
+```
+                            
+In the latest Pod definition, a field called `ephemeralContainers` has been introduced. Similar to `initContainers` and `containers`, this field contains a list of containers. However, it is important to note that `ephemeralContainers` has certain options not available in the other fields. For more details, please refer to the API documentation.
```
Comment (tengqm) [inline content/en/docs/concepts/workloads/pods/ephemeral-containers.md:115]:
```suggestion
The `ephemeralContainers` field is imilar to `initContainers` and `containers` in that
it contains a list of containers. However, the `ephemeralContainers` field has certain
options not available in the other fields. For more details, please refer to the API
documentation.
```

File: content/en/docs/concepts/workloads/pods/ephemeral-containers.md [hunk 16]
```diff
@@ -76,3 +76,261 @@ you can view processes in other containers.
...
+
+One notable feature of `ephemeralContainers` is its ability to configure the container security context. This means that in theory, a malicious actor could potentially escalate the container's privileges. However, the impact of this depends on the policy enforcement tool being used. If you are using a robust policy enforcement tool, it should be able to mitigate such risks.
```
Comment (tengqm) [inline content/en/docs/concepts/workloads/pods/ephemeral-containers.md:117]:
```suggestion
One notable feature of `ephemeralContainers` is its ability to configure the security context.
In theory, a malicious actor could escalate the container's privileges. This risk can
be effectively mitigated by a robust policy enforcement tool.
```

File: content/en/docs/concepts/workloads/pods/ephemeral-containers.md [hunk 17]
```diff
@@ -76,3 +76,261 @@ you can view processes in other containers.
...
+
+The answer to whether it affects users of policy enforcement tools varies. It depends on the specific tool and its version, as well as whether you are using policies from the project's library or custom policies developed in-house. To ensure security, it is advisable to review the policies and keep the policy enforcement tool up-to-date with the latest security measures.
```
Comment (tengqm) [inline content/en/docs/concepts/workloads/pods/ephemeral-containers.md:119]:
```suggestion
Other factors to consider include the tool version, as well as the policies enforced.
It is recommended to carefully review the policies and keep the tool up-to-date
with the latest security measures.
```

File: content/en/docs/concepts/workloads/pods/ephemeral-containers.md [hunk 18]
```diff
@@ -76,3 +76,270 @@ you can view processes in other containers.
 
 * Learn how to [debug pods using ephemeral containers](/docs/tasks/debug/debug-application/debug-running-pod/#ephemeral-container).
 
+### Kubernetes Ephemeral Container Security
```
Comment (verb) [inline content/en/docs/concepts/workloads/pods/ephemeral-containers.md:79]: This is indented under "What's Next" but should probably be its own top level heading,

```
## Ephemeral container security
### Pod Security Policies
### RBAC
etc.
```

File: content/en/docs/concepts/workloads/pods/ephemeral-containers.md [hunk 19]
```diff
@@ -76,3 +76,270 @@ you can view processes in other containers.
...
+```
+                            
+The `ephemeralContainers` field is imilar to `initContainers` and `containers` in that
+it contains a list of containers. However, the `ephemeralContainers` field has certain
+options not available in the other fields. For more details, please refer to the API
+documentation.
+
+One notable feature of `ephemeralContainers` is its ability to configure the security context.
+In theory, a malicious actor could escalate the container's privileges. This risk can
+be effectively mitigated by a robust policy enforcement tool.
```
Comment (verb) [inline content/en/docs/concepts/workloads/pods/ephemeral-containers.md:124]: Optional: [Pod Security Admission](https://kubernetes.io/docs/concepts/security/pod-security-admission/) covers ephemeral containers, if you want to point to that.

File: content/en/docs/concepts/workloads/pods/ephemeral-containers.md [hunk 20]
```diff
@@ -76,3 +76,267 @@ you can view processes in other containers.
...
+
+## RBAC
+
+Disallowing ephemeral containers with RBAC could be an option if
+the feature is not needed and it is not possible to disable the
+feature completely. The KEP-277: Ephemeral Containers state the
+following about using RBAC to disable ephemeral containers.
+
+Cluster administrators will be expected to choose from one of the
+following mechanisms for restricting usage of ephemeral containers:
```
[the remark this answers is not in the document]
Comment (verb) [reply content/en/docs/concepts/workloads/pods/ephemeral-containers.md:188]: Is it possible to format this as a quote in markdown?

Either way it would be good to link to the source, https://git.k8s.io/enhancements/keps/sig-node/277-ephemeral-containers#configurable-security-policy

File: content/en/docs/concepts/workloads/pods/ephemeral-containers.md [hunk 21]
```diff
@@ -76,3 +76,270 @@ you can view processes in other containers.
...
+
+Use RBAC to control which users are allowed to access the
+/ephemeralcontainers subresource.
+Write or use a third-party admission controller to allow 
+or reject Pod updates that modify ephemeral containers based
+on the content of the update.
+Disable the feature using the EphemeralContainers feature gate.
+RBAC is additive which means that it is not possible to remove
+permissions from a role. This type of mitigation obviously does
+not matter if all users a cluster admin, which they should not
```
Comment (verb) [inline content/en/docs/concepts/workloads/pods/ephemeral-containers.md:198]: "all users are cluster admins"

**Review by verb — COMMENTED** [review]
Thanks for adding to the ephemeral containers documentation! I see that you're actively working on this, so I'll just add a few suggestions and wait for your next revision.

File: content/en/docs/concepts/workloads/pods/ephemeral-containers.md (hunk 20 above)
[the remark this answers is not in the document]
Comment (sftim) [reply content/en/docs/concepts/workloads/pods/ephemeral-containers.md:188]: Even better: rewrite the quoted text to match our docs [style guide](https://kubernetes.io/docs/contribute/style/style-guide/).

“As a cluster operator, you are expected to choose from one of” for example.

File: content/en/docs/concepts/workloads/pods/ephemeral-containers.md [hunk 22]
```diff
@@ -76,3 +76,270 @@ you can view processes in other containers.
...
+happenstance. Either way it is good to trust but verify that
+it is not for example possible to create a privileged ephemeral 
+container. Annoyingly the debug command does not expose any options
+to set any security context configuration, so we need another option.
+Ephemeral containers cannot be defined in a Pod when it is created 
+and it can neither be added with an update. We need some other
+method to create these specific ephemeral containers.
+
+Ephemeral containers are created using a special ephemeralcontainers 
+handler in the API rather than by adding them directly to pod.spec,
```
Comment (sftim) [inline content/en/docs/concepts/workloads/pods/ephemeral-containers.md:242]:
```suggestion
You create an ephemeral container using a special `ephemeralcontainers`
subresource of the Pod API (rather than by adding them directly to the `.spec` of a Pod),
```

File: content/en/docs/concepts/workloads/pods/ephemeral-containers.md [hunk 23]
```diff
@@ -76,3 +76,270 @@ you can view processes in other containers.
...
+so it's not possible to add an ephemeral container using kubectl edit.
```
Comment (sftim) [inline content/en/docs/concepts/workloads/pods/ephemeral-containers.md:243]: Is this right?
```suggestion
so it's not possible to add (or remove) an ephemeral container using `kubectl edit`.
```

File: content/en/docs/concepts/workloads/pods/ephemeral-containers.md (hunk 18 above)
Comment (sftim) [inline content/en/docs/concepts/workloads/pods/ephemeral-containers.md:79]: I would move this into [an existing] page within https://kubernetes.io/docs/concepts/security/ and then have a _link_ to that new section.

File: content/en/docs/concepts/workloads/pods/ephemeral-containers.md [hunk 24]
```diff
@@ -76,3 +76,270 @@ you can view processes in other containers.
...
+container can enjoy all the freedom which writing files entails.
+
+Now that you have the  feature you can start a ephemeral container in any Pod you like  
+
+```shell
+kubectl run ephemeral-demo --image=k8s.gcr.io/pause:3.1 --restart=Never
+kubectl debug -it ephemeral-demo --image=busybox:1.28
+```
+
+Once you gain access to a Shell, you can perform management operations in it.
```
Comment (sftim) [inline content/en/docs/concepts/workloads/pods/ephemeral-containers.md:98]: Is this right?
```suggestion
Once you're running a shell inside the ephemeral container, you can try some troubleshooting.
You can do debug operations even if the main container doesn't include a shell (for example: 
the `pause` container image doesn't include a shell binary)
```

File: content/en/docs/concepts/workloads/pods/ephemeral-containers.md [hunk 25]
```diff
@@ -76,3 +76,270 @@ you can view processes in other containers.
...
+metadata:
+  name: ephemeral-demo
+spec:
+  ...
+  ephemeralContainers:
+  - name: debugger-r59b7
+    image: busybox:1.28
+    imagePullPolicy: IfNotPresent
+    stdin: true
+    tty: true
```
Comment (sftim) [inline content/en/docs/concepts/workloads/pods/ephemeral-containers.md:114]: Does this work as a manifest?

**Review by sftim — COMMENTED** [review]
@deepak4566 have you decided whether you want to write documentation or a blog article? If you've decided: which did you pick?

File: content/en/docs/concepts/workloads/pods/ephemeral-containers.md (hunk 23 above)
Comment (verb) [reply content/en/docs/concepts/workloads/pods/ephemeral-containers.md:243]: Yes, but also it's not possible to remove ephemeral containers by any method.

Comment (sftim) [reply content/en/docs/concepts/workloads/pods/ephemeral-containers.md:243]: Let's document that as a thing to watch out for: once an ephemeral container is there, you're stuck with it.

**Comment by sftim** [comment]
@deepak4566 you are welcome to revise this per https://github.com/kubernetes/website/pull/42357/files#r1282923297

If you'd like guidance you can ask in #sig-docs on the Kubernetes Slack workspace, to see if anyone has capacity to help you learn about technical writing.

I can also recommend the book [Docs for Developers](https://docsfordevelopers.com/), co-written by several eminent technical writers. Amongst the authors are two of our own emeritus approvers, Sarah Corleissen and Jared Bhatti.

**Comment by divya-mohan0209** [comment]
Hello @deepak4566 , thank you for your contribution. Since there has been no activity on this PR for quite some time (~4 months), I will be closing this PR. Tim has added valuable feedback to assist you in your technical writing journey and we hope that it is useful. For guidelines specific to the documentation of the Kubernetes project, please feel free to refer to the [contributor guide](https://kubernetes.io/docs/contribute/). We look forward to seeing you contribute :)

/close

**Comment by deepak4566** [comment]
> Hello @deepak4566 , thank you for your contribution. Since there has been no activity on this PR for quite some time (~4 months), I will be closing this PR. Tim has added valuable feedback to assist you in your technical writing journey and we hope that it is useful. For guidelines specific to the documentation of the Kubernetes project, please feel free to refer to the [contributor guide](https://kubernetes.io/docs/contribute/). We look forward to seeing you contribute :)
> 
> /close

Hey , i wasn inactive in github due to some of university work i want to reopen this pr
